datafusion/execution/src/dedicated_executor.rs

```diff
@@ -0,0 +1,1004 @@
+    /// pool managed by this `DedicatedExecutor`.
+    ///
+    /// # TODO: make this wait (aka so the API doesn't start a new background task or whatever)
+    ///
+    /// # Notes
+    ///
+    /// This task is run on a dedicated Tokio runtime that purposely does not have
+    /// IO enabled. If your future makes any IO calls, you have to
+    /// explicitly run them on DedicatedExecutor::spawn_io.
+    ///
+    /// If you see a message like this
+    ///
+    /// (Panic { msg: "A Tokio 1.x context was found, but timers are disabled. Call `enable_time` on the runtime builder to enable timers."
+    ///
+    /// It means some work that was meant to be done on the IO runtime was done
+    /// on the CPU runtime.
+    ///
+    /// UNLIKE [`tokio::task::spawn`], the returned future is **cancelled** when
+    /// it is dropped. Thus, you need ensure the returned future lives until it
+    /// completes (call `await`) or you wish to cancel it.
+    ///
+    /// All spawned tasks are added to the tokio executor immediately and
+    /// compete for the threadpool's resources.
+    pub fn spawn_cpu<T>(
+        &self,
+        task: T,
+    ) -> impl Future<Output = Result<T::Output, JobError>>
+    where
+        T: Future + Send + 'static,
+        T::Output: Send + 'static,
+    {
+        let handle = {
+            let state = self.state.read();
+            state.handle.clone()
+        };
+
+        let Some(handle) = handle else {
+            return futures::future::err(JobError::WorkerGone).boxed();
+        };
+
+        // use JoinSet implement "cancel on drop"
+        let mut join_set = JoinSet::new();
+        join_set.spawn_on(task, &handle);
+        async move {
+            join_set
+                .join_next()
+                .await
+                .expect("just spawned task")
+                .map_err(|e| match e.try_into_panic() {
+                    Ok(e) => {
+                        let s = if let Some(s) = e.downcast_ref::<String>() {
+                            s.clone()
+                        } else if let Some(s) = e.downcast_ref::<&str>() {
+                            s.to_string()
+                        } else {
+                            "unknown internal error".to_string()
+                        };
+
+                        JobError::Panic { msg: s }
+                    }
+                    Err(_) => JobError::WorkerGone,
+                })
+        }
+        .boxed()
+    }
+
+    /// Runs the specified work on the dedicated executor and returns the result
+    ///
+    /// Note the future is not 'static (aka it can have internal references)
+    pub fn spawn_cpu2<'a, T>(&self, task: T) -> impl Future<Output = T::Output>
+    where
+        T: Future + Send + 'a,
+        T::Output: Send,
+    {
+        // If we can figure out how to make this work, then
+        // we could integrate it nicely into DataFusion
+        async { todo!() }
+    }
```

Comment:
If we can figure out  way to implement this function, the "annotate all callsites" is plausible

The key difference is:
```diff
-        T: Future + Send + 'static,
+        T: Future + Send + 'a,
```

The idea is there are a bunch of async function calls like this:

```rust
                self
                    .list_files_for_scan(
                        &session_state,
                        &partition_filters,
                        limit,
                    )
                    .await
```

Rust can figure out somehow that if this future runs on the same runtime, the references are valid for long enough and will compile it.

However when I tried to run this kind of function on a different runtime (the whole point of this PR), Rust can't figure out that the future body doesn't escape the current function and thus I can't get it to compile. The only way to pass the future to a different runtime is if the lifetime is `'static`. 

I believe the root cause is that all the functions for spawning / running tasks on a tokio runtime require the Futures are `'static` -- that is that they have no references internally. For example:
- https://docs.rs/tokio/latest/tokio/task/fn.spawn.html
![Screenshot 2024-12-08 at 12 03 52 PM](https://github.com/user-attachments/assets/745ff9ca-5e23-4d1b-a4c2-96927bc8414e)


The only way I could make it work was to clone all the arguments (we could likely avoid this by restructing the code to be much more careful about what arguments are passed, need to be cloned, etc) but doing so would be a pretty large undertaking and likely require significantly more copying, like:


```rust
        let self_captures = self.clone(); // clone so we can pass future to the other pool
        let (mut partitioned_file_lists, statistics) = state
            .runtime_env()
            .spawn_io(async move {
                self_captured
                    .list_files_for_scan(
                        &session_state_captured,
                        &partition_filters_captured,
                        limit,
                    )
                    .await
            })
            .await?;
```

Reply:
I don't think the lifetimes you propose are possible for the reasons you've articulated, tokio doesn't provide structured concurrency.

> but doing so would be a pretty large undertaking and likely require significantly more copying, like:

In most cases this should be a case of extracting the shared IO state onto a separate struct that can be Arc-wrapped. Whilst this is work, separating the IO component explicitly in this way is IMO actually a good idea anyway

Reply:
> In most cases this should be a case of extracting the shared IO state onto a separate struct that can be Arc-wrapped. Whilst this is work, separating the IO component explicitly in this way is IMO actually a good idea anyway

That seems reasonable. I think the challenge is going to be restructuring the various `ListingTable`  APIs in this way. 

i think if we can show how to make it work with `ListingTable` we can make the approach work for anything else in DataFusion's codebase